frontend/packages/topology/src/layouts/DagreLayout.ts

```diff
@@ -1,119 +1,142 @@
 import * as dagre from 'dagre';
 import * as _ from 'lodash';
-import { Edge, GraphElement, Graph, Layout, Node } from '../types';
+import { Edge, Graph, Layout, Node } from '../types';
 import Point from '../geom/Point';
-import { leafNodeElements } from '../utils/element-utils';
-
-class DagreNode {
-  private node: Node;
-
-  constructor(node: Node) {
-    this.node = node;
-  }
-
-  getId(): string {
-    return this.node.getId();
+import { BaseLayout, LayoutGroup, LayoutLink, LayoutNode, LayoutOptions } from './BaseLayout';
+
+class DagreNode extends LayoutNode implements dagre.Node {
+  getUpdatableNode(): dagre.Node {
+    return {
+      id: this.id,
+      width: this.width,
+      height: this.height,
+      x: this.x,
+      y: this.y,
+    };
   }
 
-  getData(): any {
-    return this.node.getData();
+  updateToNode(updatedNode: dagre.Node | undefined): void {
+    if (updatedNode) {
+      this.x = updatedNode.x;
+      this.y = updatedNode.y;
+      this.update();
+    }
   }
+}
 
-  get element(): Node {
-    return this.node;
-  }
+class DagreGroup extends LayoutGroup {}
 
-  get x(): number {
-    return this.node.getBounds().getCenter().x;
-  }
+class DagreLink extends LayoutLink {
+  public points: any[];
 
-  set x(value: number) {
-    this.node.setBounds(
-      this.node.getBounds().setCenter(value, this.node.getBounds().getCenter().y),
-    );
+  updateBendpoints(): void {
+    if (this.points && !this.isFalse && this.points.length > 2) {
+      this.element.setBendpoints(
+        this.points.slice(1, -1).map((point: any) => new Point(point.x, point.y)),
+      );
+    }
   }
+}
 
-  get y(): number {
-    return this.node.getBounds().getCenter().y;
+type DagreLayoutOptions = LayoutOptions & {
```

Comment:
```suggestion
type DagreLayoutOptions = LayoutOptions & dagre.GraphLabel;
```
`dagre.GraphLabel` is all the options that the graph can take.